File: src/main/java/net/lumalyte/lumasg/util/ScoreboardCache.java
package net.lumalyte.util;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import net.lumalyte.LumaSG;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Scoreboard;

import java.time.Duration;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * High-performance scoreboard caching system
 * Reduces scoreboard computation overhead and improves update performance
 */
public class ScoreboardCache {
    
    // Cache for scoreboard line content to avoid string concatenation overhead
    private static final Cache<String, String> SCOREBOARD_LINE_CACHE = Caffeine.newBuilder()
            .maximumSize(1000)
            .expireAfterWrite(Duration.ofSeconds(30))
            .recordStats()
            .build();
    
    // Cache for computed placeholder values
    private static final Cache<String, String> PLACEHOLDER_CACHE = Caffeine.newBuilder()
            .maximumSize(500)
            .expireAfterWrite(Duration.ofSeconds(10))
            .recordStats()
            .build();
    
    // Cache for formatted time strings
    private static final Cache<Integer, String> TIME_FORMAT_CACHE = Caffeine.newBuilder()
            .maximumSize(3600) // Cache up to 1 hour of time formats
            .expireAfterWrite(Duration.ofMinutes(5))
            .recordStats()
            .build();
    
    // Cache for player-specific scoreboard data
    private static final Map<UUID, PlayerScoreboardData> PLAYER_SCOREBOARD_DATA = new ConcurrentHashMap<>();
    
    private static LumaSG plugin;
    private static DebugLogger.ContextualLogger logger;
    
    /**
     * Player-specific scoreboard data container
     */
    private static class PlayerScoreboardData {
        private final Map<String, String> cachedLines = new ConcurrentHashMap<>();
        private long lastUpdate = 0;
        private boolean needsRefresh = true;
        
        public void setCachedLine(String key, String value) {
            cachedLines.put(key, value);
        }
        
        public String getCachedLine(String key) {
            return cachedLines.get(key);
        }
        
        public void markForRefresh() {
            needsRefresh = true;
            lastUpdate = System.currentTimeMillis();
        }
        
        public boolean needsRefresh(long updateInterval) {
            return needsRefresh || (System.currentTimeMillis() - lastUpdate) > updateInterval;
        }
        
        public void refreshCompleted() {
            needsRefresh = false;
            lastUpdate = System.currentTimeMillis();
        }
    }
    
    /**
     * Initializes the scoreboard cache
     * 
     * @param pluginInstance The plugin instance
     */
    public static void initialize(LumaSG pluginInstance) {
        plugin = pluginInstance;
        logger = plugin.getDebugLogger().forContext("ScoreboardCache");
        
        logger.info("Scoreboard Cache initialized");
    }
    
    /**
     * Gets cached scoreboard line content
     * 
     * @param lineKey The unique key for the line
     * @param generator The function to generate the line if not cached
     * @return The cached or generated line content
     */
    public static String getCachedScoreboardLine(String lineKey, java.util.function.Supplier<String> generator) {
        return SCOREBOARD_LINE_CACHE.get(lineKey, key -> generator.get());
    }
    
    /**
     * Gets cached placeholder value
     * 
     * @param placeholderKey The placeholder key
     * @param generator The function to generate the value if not cached
     * @return The cached or generated placeholder value
     */
    public static String getCachedPlaceholder(String placeholderKey, java.util.function.Supplier<String> generator) {
        return PLACEHOLDER_CACHE.get(placeholderKey, key -> generator.get());
    }
    
    /**
     * Gets cached formatted time string
     * 
     * @param timeInSeconds The time in seconds
     * @return The formatted time string
     */
    public static String getCachedTimeFormat(int timeInSeconds) {
        return TIME_FORMAT_CACHE.get(timeInSeconds, time -> {
            int minutes = time / 60;
            int seconds = time % 60;
            return String.format("%02d:%02d", minutes, seconds);
        });
    }
    
    /**
     * Gets or creates player scoreboard data
     * 
     * @param playerId The player's UUID
     * @return The player's scoreboard data
     */
    public static PlayerScoreboardData getPlayerScoreboardData(UUID playerId) {
        return PLAYER_SCOREBOARD_DATA.computeIfAbsent(playerId, id -> new PlayerScoreboardData());
    }
    
    /**
     * Checks if a player's scoreboard needs updating
     * 
     * @param playerId The player's UUID
     * @param updateInterval The update interval in milliseconds
     * @return True if the scoreboard needs updating
     */
    public static boolean needsScoreboardUpdate(UUID playerId, long updateInterval) {
        PlayerScoreboardData data = getPlayerScoreboardData(playerId);
        return data.needsRefresh(updateInterval);
    }
    
    /**
     * Marks a player's scoreboard as needing refresh
     * 
     * @param playerId The player's UUID
     */
    public static void markScoreboardForRefresh(UUID playerId) {
        PlayerScoreboardData data = getPlayerScoreboardData(playerId);
        data.markForRefresh();
    }
    
    /**
     * Marks a player's scoreboard refresh as completed
     * 
     * @param playerId The player's UUID
     */
    public static void markScoreboardRefreshCompleted(UUID playerId) {
        PlayerScoreboardData data = getPlayerScoreboardData(playerId);
        data.refreshCompleted();
    }
    
    /**
     * Caches a scoreboard line for a specific player
     * 
     * @param playerId The player's UUID
     * @param lineKey The line key
     * @param lineContent The line content
     */
    public static void cachePlayerScoreboardLine(UUID playerId, String lineKey, String lineContent) {
        PlayerScoreboardData data = getPlayerScoreboardData(playerId);
        data.setCachedLine(lineKey, lineContent);
    }
    
    /**
     * Gets a cached scoreboard line for a specific player
     * 
     * @param playerId The player's UUID
     * @param lineKey The line key
     * @return The cached line content, or null if not cached
     */
    public static String getCachedPlayerScoreboardLine(UUID playerId, String lineKey) {
        PlayerScoreboardData data = PLAYER_SCOREBOARD_DATA.get(playerId);
        return data != null ? data.getCachedLine(lineKey) : null;
    }
    
    /**
     * Invalidates all cached data for a player
     * 
     * @param playerId The player's UUID
     */
    public static void invalidatePlayer(UUID playerId) {
        PLAYER_SCOREBOARD_DATA.remove(playerId);
        
        // Remove player-specific entries from global caches
        String playerPrefix = playerId.toString();
        SCOREBOARD_LINE_CACHE.asMap().keySet().removeIf(key -> key.startsWith(playerPrefix));
        PLACEHOLDER_CACHE.asMap().keySet().removeIf(key -> key.startsWith(playerPrefix));
    }
    
    /**
     * Invalidates cached placeholders
     * 
     * @param placeholderPrefix The prefix of placeholders to invalidate
     */
    public static void invalidatePlaceholders(String placeholderPrefix) {
        PLACEHOLDER_CACHE.asMap().keySet().removeIf(key -> key.startsWith(placeholderPrefix));
    }
    
    /**
     * Gets comprehensive cache statistics
     * 
     * @return String containing cache statistics
     */
    public static String getCacheStats() {
        return String.format(
            "Scoreboard Cache Stats:\n" +
            "Scoreboard Lines - Size: %d, Hit Rate: %.2f%%\n" +
            "Placeholders - Size: %d, Hit Rate: %.2f%%\n" +
            "Time Formats - Size: %d, Hit Rate: %.2f%%\n" +
            "Player Data Entries: %d",
            SCOREBOARD_LINE_CACHE.estimatedSize(), SCOREBOARD_LINE_CACHE.stats().hitRate() * 100,
            PLACEHOLDER_CACHE.estimatedSize(), PLACEHOLDER_CACHE.stats().hitRate() * 100,
            TIME_FORMAT_CACHE.estimatedSize(), TIME_FORMAT_CACHE.stats().hitRate() * 100,
            PLAYER_SCOREBOARD_DATA.size()
        );
    }
    
    /**
     * Performs cache maintenance
     */
    public static void performMaintenance() {
        SCOREBOARD_LINE_CACHE.cleanUp();
        PLACEHOLDER_CACHE.cleanUp();
        TIME_FORMAT_CACHE.cleanUp();
        
        // Clean up inactive player data
        long cutoffTime = System.currentTimeMillis() - Duration.ofMinutes(15).toMillis();
        PLAYER_SCOREBOARD_DATA.entrySet().removeIf(entry -> entry.getValue().lastUpdate < cutoffTime);
        
        logger.debug("Scoreboard cache maintenance completed");
    }
    
    /**
     * Clears all caches
     */
    public static void clearAllCaches() {
        SCOREBOARD_LINE_CACHE.invalidateAll();
        PLACEHOLDER_CACHE.invalidateAll();
        TIME_FORMAT_CACHE.invalidateAll();
        PLAYER_SCOREBOARD_DATA.clear();
        
        logger.info("All scoreboard caches cleared");
    }
    
    /**
     * Creates optimized scoreboard line with caching
     * 
     * @param gameId The game ID
     * @param lineType The type of line (players, time, arena, etc.)
     * @param value The value to display
     * @return The formatted scoreboard line
     */
    public static String createOptimizedScoreboardLine(String gameId, String lineType, String value) {
        String cacheKey = gameId + ":" + lineType + ":" + value;
        
        return getCachedScoreboardLine(cacheKey, () -> {
            return switch (lineType) {
                case "players" -> "§f§lPlayers: §e" + value;
                case "time" -> "§f§lTime: §c" + value;
                case "arena" -> "§f§lArena: §b" + value;
                case "alive" -> "§f§lAlive: §a" + value;
                default -> "§f" + value;
            };
        });
    }
    
    /**
     * Bulk updates scoreboard for multiple players efficiently
     * 
     * @param players The players to update
     * @param lineUpdates The line updates to apply
     */
    public static void bulkUpdateScoreboards(Iterable<Player> players, Map<String, String> lineUpdates) {
        for (Player player : players) {
            UUID playerId = player.getUniqueId();
            PlayerScoreboardData data = getPlayerScoreboardData(playerId);
            
            // Check if any lines actually changed
            boolean hasChanges = false;
            for (Map.Entry<String, String> update : lineUpdates.entrySet()) {
                String cachedLine = data.getCachedLine(update.getKey());
                if (!update.getValue().equals(cachedLine)) {
                    hasChanges = true;
                    break;
                }
            }
            
            // Only update if there are actual changes
            if (hasChanges) {
                for (Map.Entry<String, String> update : lineUpdates.entrySet()) {
                    data.setCachedLine(update.getKey(), update.getValue());
                }
                data.markForRefresh();
            }
        }
    }
} 

File: src/main/kotlin/net/lumalyte/lumasg/util/cache/ScoreboardCache.kt
package net.lumalyte.lumasg.util.cache

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import net.badgersmc.nexus.annotations.Service
import org.bukkit.entity.Player
import org.slf4j.LoggerFactory
import java.time.Duration
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * High-performance scoreboard caching system.
 * Reduces scoreboard computation overhead and improves update performance.
 */
@Service
class ScoreboardCache {

    private val logger = LoggerFactory.getLogger(ScoreboardCache::class.java)

    /** Cache for scoreboard line content to avoid string concatenation overhead. */
    private val scoreboardLineCache: Cache<String, String> = Caffeine.newBuilder()
        .maximumSize(1000)
        .expireAfterWrite(Duration.ofSeconds(30))
        .recordStats()
        .build()

    /** Cache for computed placeholder values. */
    private val placeholderCache: Cache<String, String> = Caffeine.newBuilder()
        .maximumSize(500)
        .expireAfterWrite(Duration.ofSeconds(10))
        .recordStats()
        .build()

    /** Cache for formatted time strings (up to 1 hour of formats). */
    private val timeFormatCache: Cache<Int, String> = Caffeine.newBuilder()
        .maximumSize(3600)
        .expireAfterWrite(Duration.ofMinutes(5))
        .recordStats()
        .build()

    /** Per-player scoreboard data. */
    private val playerScoreboardData = ConcurrentHashMap<UUID, PlayerScoreboardData>()

    /** Player-specific scoreboard data container. */
    class PlayerScoreboardData {
        val cachedLines = ConcurrentHashMap<String, String>()
        var lastUpdate: Long = 0L
            private set
        var needsRefresh: Boolean = true
            private set

        fun setCachedLine(key: String, value: String) {
            cachedLines[key] = value
        }

        fun getCachedLine(key: String): String? = cachedLines[key]

        fun markForRefresh() {
            needsRefresh = true
            lastUpdate = System.currentTimeMillis()
        }

        fun needsRefresh(updateInterval: Long): Boolean =
            needsRefresh || (System.currentTimeMillis() - lastUpdate) > updateInterval

        fun refreshCompleted() {
            needsRefresh = false
            lastUpdate = System.currentTimeMillis()
        }
    }

    // ── Public API ──────────────────────────────────────────────────────────

    /** Gets cached scoreboard line content, generating via [generator] if absent. */
    fun getCachedScoreboardLine(key: String, generator: () -> String): String =
        scoreboardLineCache.get(key) { generator() }!!

    /** Gets cached placeholder value, generating via [generator] if absent. */
    fun getCachedPlaceholder(key: String, generator: () -> String): String =
        placeholderCache.get(key) { generator() }!!

    /** Gets a cached formatted time string (MM:SS). */
    fun getCachedTimeFormat(timeInSeconds: Int): String =
        timeFormatCache.get(timeInSeconds) { time ->
            val minutes = time / 60
            val seconds = time % 60
            "%02d:%02d".format(minutes, seconds)
        }!!

    /** Checks if a player's scoreboard needs updating based on the interval. */
    fun needsScoreboardUpdate(playerId: UUID, updateInterval: Long): Boolean =
        getOrCreatePlayerData(playerId).needsRefresh(updateInterval)

    /** Marks a player's scoreboard as needing refresh. */
    fun markScoreboardForRefresh(playerId: UUID) {
        getOrCreatePlayerData(playerId).markForRefresh()
    }

    /** Marks a player's scoreboard refresh as completed. */
    fun markScoreboardRefreshCompleted(playerId: UUID) {
        getOrCreatePlayerData(playerId).refreshCompleted()
    }

    /** Caches a scoreboard line for a specific player. */
    fun cachePlayerScoreboardLine(playerId: UUID, lineKey: String, lineContent: String) {
        getOrCreatePlayerData(playerId).setCachedLine(lineKey, lineContent)
    }

    /** Gets a cached scoreboard line for a specific player, or null if not cached. */
    fun getCachedPlayerScoreboardLine(playerId: UUID, lineKey: String): String? =
        playerScoreboardData[playerId]?.getCachedLine(lineKey)

    // ── Invalidation ────────────────────────────────────────────────────────

    /** Invalidates all cached data for a player. */
    fun invalidatePlayer(playerId: UUID) {
        playerScoreboardData.remove(playerId)

        val playerPrefix = playerId.toString()
        scoreboardLineCache.asMap().keys.removeIf { it.startsWith(playerPrefix) }
        placeholderCache.asMap().keys.removeIf { it.startsWith(playerPrefix) }
    }

    /** Invalidates cached placeholders matching a prefix. */
    fun invalidatePlaceholders(prefix: String) {
        placeholderCache.asMap().keys.removeIf { it.startsWith(prefix) }
    }

    // ── Optimized helpers ───────────────────────────────────────────────────

    /** Creates an optimized, cached scoreboard line for a game context. */
    fun createOptimizedScoreboardLine(gameId: String, lineType: String, value: String): String {
        val cacheKey = "$gameId:$lineType:$value"
        return getCachedScoreboardLine(cacheKey) {
            when (lineType) {
                "players" -> "\u00a7f\u00a7lPlayers: \u00a7e$value"
                "time" -> "\u00a7f\u00a7lTime: \u00a7c$value"
                "arena" -> "\u00a7f\u00a7lArena: \u00a7b$value"
                "alive" -> "\u00a7f\u00a7lAlive: \u00a7a$value"
                else -> "\u00a7f$value"
            }
        }
    }

    /** Bulk updates scoreboards for multiple players, only refreshing when lines actually changed. */
    fun bulkUpdateScoreboards(players: Iterable<Player>, lineUpdates: Map<String, String>) {
        for (player in players) {
            val playerId = player.uniqueId
            val data = getOrCreatePlayerData(playerId)

            val hasChanges = lineUpdates.any { (key, value) ->
                data.getCachedLine(key) != value
            }

            if (hasChanges) {
                lineUpdates.forEach { (key, value) -> data.setCachedLine(key, value) }
                data.markForRefresh()
            }
        }
    }

    // ── Maintenance ─────────────────────────────────────────────────────────

    /** Clears all caches and player data. */
    fun clearAllCaches() {
        scoreboardLineCache.invalidateAll()
        placeholderCache.invalidateAll()
        timeFormatCache.invalidateAll()
        playerScoreboardData.clear()
        logger.info("All scoreboard caches cleared")
    }

    /** Returns comprehensive cache statistics. */
    fun getCacheStats(): String {
        val lineStats = scoreboardLineCache.stats()
        val phStats = placeholderCache.stats()
        val tfStats = timeFormatCache.stats()

        return """
            Scoreboard Cache Stats:
            Scoreboard Lines - Size: ${scoreboardLineCache.estimatedSize()}, Hit Rate: ${"%.2f".format(lineStats.hitRate() * 100)}%
            Placeholders - Size: ${placeholderCache.estimatedSize()}, Hit Rate: ${"%.2f".format(phStats.hitRate() * 100)}%
            Time Formats - Size: ${timeFormatCache.estimatedSize()}, Hit Rate: ${"%.2f".format(tfStats.hitRate() * 100)}%
            Player Data Entries: ${playerScoreboardData.size}
        """.trimIndent()
    }

    /** Performs cache maintenance: evicts expired entries and cleans up inactive players. */
    fun performMaintenance() {
        scoreboardLineCache.cleanUp()
        placeholderCache.cleanUp()
        timeFormatCache.cleanUp()

        // Clean up inactive player data (older than 15 minutes)
        val cutoffTime = System.currentTimeMillis() - Duration.ofMinutes(15).toMillis()
        playerScoreboardData.entries.removeIf { it.value.lastUpdate < cutoffTime }

        logger.debug("Scoreboard cache maintenance completed")
    }

    // ── Internal ────────────────────────────────────────────────────────────

    private fun getOrCreatePlayerData(playerId: UUID): PlayerScoreboardData =
        playerScoreboardData.computeIfAbsent(playerId) { PlayerScoreboardData() }
}
